pkg/kubelet/kubelet.go

```diff
@@ -1739,8 +1749,8 @@ func (kl *Kubelet) syncPod(pod *api.Pod, mirrorPod *api.Pod, podStatus *kubecont
 	}
 	kl.statusManager.SetPodStatus(pod, apiPodStatus)
 
-	// Kill pods we can't run.
-	if err := canRunPod(pod); err != nil || pod.DeletionTimestamp != nil {
+	// Kill pods we can't run, pods pending deletion, or pods that have been marked as failed.
```

Comment:
I am not sure I follow the major issue with this race condition.  In either case, the end result is the desired result and the pod is killed, and HandlePodCleanup will forget the worker, no?  Not sure if you are asking for something additional in this PR.  Let me know.